Add support to delete helm releases on failure (#963)
hmcts/cnp-jenkins-library: README.md

The API tests run after smoke tests.

#### Clear Helm Release

> If your server never uses the deployed resources, teams can clear the helm release to free resources on the cluster.

Reviewer: ```suggestion
If your service never uses the deployed resources after the pipeline is successful or if you refer to them only for logs in case of failure, you can uninstall the helm release to free resources on the cluster. Note that clearing the helm release won't deny you access to the pods logs, as they are saved as artefacts in Jenkins before the helm release is cleared.
```

> This can be done always or depending on success.

Reviewer: ```suggestion
```

The API tests run after smoke tests.

#### Clear Helm Release

If your server never uses the deployed resources, teams can clear the helm release to free resources on the cluster.
This can be done always or depending on success.

> To note that clearing the helm release won't deny you access to the pods logs, as they are saved before the helm release is cleared.

Reviewer: ```suggestion
```

  }
  ```

> #### Clear Helm Release on Failed Build

Reviewer 2: is there a reason we wouldn't want to do this by default given we're capturing the pods logs? I guess if you wanted to check the pod environment variables etc? but we should capture the pod manifest as well maybe ?

Author: The reason is simply that I don't presume to know how other teams like to work.

For now it's just safer to add the feature as an "opt-in". Then platops can do further considerations as to the opportunity to enable this by default (which I would imagine would be the best thing to do) and what other changes are needed to support that properly.

If your team doesn't need to keep the helm releases for their PR even for failing builds, it is possible to clean them to free further resources.
This may be useful to reduce resource contention in difficult projects with large teams, where pipeline stability is not yet where it should be.

To clear helm releases for failing builds, do the following:

  ```
  withPipeline(type, product, component) {
    ...

>     enableCleanupOfHelmReleaseOnFail()

Reviewer 2: It would be good if there was a way so that you don't need to add two lines to cleanup success and failure, i.e. in your screenshot you're having to add two lines of config

You could add something like this as well?
```
enableCleanupOfHelmReleaseAlways()
```

That would map to the existing callback names?
https://github.com/hmcts/cnp-jenkins-library#extending-the-opinionated-pipeline

🚫 At a minimum I would suggest renaming this to match the callbacks:

```suggestion
    enableCleanupOfHelmReleaseOnFailure()
```

Author: I'll match the callback names.

As for the additional "always" bit, I would leave further simplifications to your team. The purpose of this PR from my standpoint is simply to finally overcome an issue that is frequently blocking us in civil.

  }
  ```

To clear the helm release on a failing build, do the following:

  ```
  withPipeline(type, product, component) {
    ...
    enableCleanupOfHelmReleaseOnFailure()
    ...
  }
  ```

> To note that clearing the helm release won't deny you access to the pods logs, as they are saved before the helm release is cleared.

Reviewer: ```suggestion
```
